dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/supplier/DuplicateRelationshipSupplier.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.preheat.supplier;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.relationship.RelationshipStore;
+import org.hisp.dhis.tracker.TrackerImportParams;
+import org.hisp.dhis.tracker.domain.Relationship;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.hisp.dhis.tracker.util.RelationshipKeySupport;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class DuplicateRelationshipSupplier extends AbstractPreheatSupplier
+{
+    @NonNull
+    private final RelationshipStore relationshipStore;
+
+    @Override
+    public void preheatAdd( TrackerImportParams params, TrackerPreheat preheat )
+    {
+        List<org.hisp.dhis.relationship.Relationship> relationships = retrieveRelationshipKeys(
+            params.getRelationships() );
+
+        relationships.forEach( r -> {
+            preheat.addDuplicatedRelationship( r.getKey() );
```

Comment:
why did you move the logic out of `addDuplicatedRelationship` and in here?

I think it belongs in `addDuplicatedRelationship`. Only then can you make sure that what is stored is not just any String but a RelatiohshipKey. As well as that the inverted key is stored for bidirectional rels.

Reply:
I think that in the end I was doing like that because it was easier to test 😬 
But I agree that it should be put back in the Preheat